### PR TITLE
New logging support for Tscope

### PR DESCRIPTION
**Pull request by Bvasilchikov**
### Description

This PR adds a custom format for logs that get output by thunderscope, and adds support for future logging.

### TODO:

- [ ] Add logging argument for desired level, default to `INFO`
- [ ] Remove old logger found under `src/software/logger/`
- [ ] Replace current `logging` library with `loguru` library

### Testing Done

Never heard of such a thing

### Resolved Issues

#3165

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: src/software/thunderscope/log/thunderscope_std_out_formatter.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    color_prefix = "\x1b[3%d;"
+    bold_code = "1m"
+    flashing_code = "5m"
+    log_format_string = (
+        "%(asctime)s - %(name)s - %(levelname)s - %(message)s (%(filename)s:%(lineno)d)"
+    )
+
+    FORMATS = {
+        logging.FATAL: color_prefix % RED
+        + flashing_code
```
Comment (williamckha) [inline src/software/thunderscope/log/thunderscope_std_out_formatter.py:20]: Blinking text might be a bit distracting/hard to read

File: src/software/thunderscope/thunderscope_main.py [hunk 2]
```diff
@@ -212,6 +227,8 @@
     )
 
     # Sanity check that an interface was provided
+    # TODO: setting `--disable_communication` still requires an interface to be provided...
```
Comment (williamckha) [inline src/software/thunderscope/thunderscope_main.py:230]: Could you tag #3172 in this TODO

File: src/software/thunderscope/thunderscope_main.py [hunk 3]
```diff
@@ -38,9 +38,24 @@
...
+    ch.setLevel(logging.DEBUG)
+    ch.setFormatter(ThunderscopeStandardOutputFormatter())
+    logger.addHandler(ch)
 
+    ###########################################################################
+    #                              Argument Parsing                           #
+    ###########################################################################
+    #
     # Setup parser
+    # TODO: create set-up func's and move into `thunderscope_config.py`
```
Comment (williamckha) [inline src/software/thunderscope/thunderscope_main.py:58]: Is this TODO related to this PR?

File: src/software/thunderscope/log/thunderscope_std_out_formatter.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+
+class ThunderscopeStandardOutputFormatter(logging.Formatter):
+    reset_code = "\x1b[0m"
+    color_prefix = "\x1b[3%d;"
+    bold_code = "1m"
+    flashing_code = "5m"
+    log_format_string = (
+        "%(asctime)s - %(name)s - %(levelname)s - %(message)s (%(filename)s:%(lineno)d)"
+    )
```
Comment (williamckha) [inline src/software/thunderscope/log/thunderscope_std_out_formatter.py:16]: Constants should be in uppercase

File: src/software/thunderscope/log/thunderscope_std_out_formatter.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
+import logging
+
+# ANSI escape codes can be found at this answer:
+# https://stackoverflow.com/questions/4842424/list-of-ansi-color-escape-sequences
+
+BLACK, RED, GREEN, YELLOW, BLUE, MAGENTA, CYAN, GREY = range(8)
```
Comment (williamckha) [inline src/software/thunderscope/log/thunderscope_std_out_formatter.py:6]: Could you turn this into an enum? e.g.
```python
Color = IntEnum("Color", ["BLACK", "RED", "GREEN", "YELLOW", "BLUE", "MAGENTA", "CYAN", "GREY"])
```

**Review by williamckha — COMMENTED** [review]
Great change! Definitely a good idea to use a proper logger in Python

File: src/software/thunderscope/thunderscope_main.py [hunk 6]
```diff
@@ -38,9 +38,24 @@
...
 if __name__ == "__main__":
 
-    logging.getLogger().setLevel(logging.INFO)
+    ###########################################################################
+    #                              Logging                                    #
+    ###########################################################################
+    #
+    # Setup logging
+    logger = logging.getLogger("Thunderscope")
+    logger.setLevel(logging.DEBUG)
```
Comment (sauravbanna) [inline src/software/thunderscope/thunderscope_main.py:47]: do we wanna change from logging.INFO to logging.DEBUG?

**Review by sauravbanna — CHANGES_REQUESTED** [review]
LGTM pending comments

File: src/software/thunderscope/log/thunderscope_std_out_formatter.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+
+class ThunderscopeStandardOutputFormatter(logging.Formatter):
```
Comment (itsarune) [inline src/software/thunderscope/log/thunderscope_std_out_formatter.py:9]: missing docs in this class

File: src/software/thunderscope/thunderscope_main.py (hunk 6 above)
Comment (itsarune) [reply src/software/thunderscope/thunderscope_main.py:47]: could by a command line arg

File: src/software/thunderscope/thunderscope_main.py (hunk 3 above)
Comment (itsarune) [reply src/software/thunderscope/thunderscope_main.py:58]: if not, add an issue

**Comment by williamckha** [comment]
Closing because PR is stale. We should definitely figure out proper logging in Python though. I think it would be a good idea to use `loguru` like Boris suggested
